src/python_testing/TC_DEM_2_9.py

```diff
@@ -0,0 +1,103 @@
+        return "4.1.3. [TC-DEM-2.2] Power or State Forecast Reporting feature functionality with DUT as Server"
+
+    def pics_TC_DEM_2_9(self):
+        """Return the PICS definitions associated with this test."""
+        pics = [
+            # Depends on Feature 01 (PowerForecastReporting) | Feature 2 (StateForecastReporting)
+            "DEM.S.F01 | DEM.S.F02",
+        ]
+        return pics
+
+    def steps_TC_DEM_2_9(self) -> list[TestStep]:
+        """Execute the test steps."""
+        steps = [
+            TestStep("1", "Commissioning, already done",
+                     is_commissioning=True),
+            TestStep("2", "TH reads TestEventTriggersEnabled attribute from General Diagnostics Cluster.",
+                     "Verify that TestEventTriggersEnabled attribute has a value of 1 (True)"),
+            TestStep("3", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for Forecast Test Event",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+            TestStep("3a", "TH reads Forecast attribute.",
+                     "Value has to include a valid slots[0].ManufacturerESAState"),
+            TestStep("3b", "TH reads Forecast attribute.",
+                     "Value has to include valid slots[0].NominalPower, slots[0].MinPower, slots[0].MaxPower, slots[0].NominalEnergy"),
+            TestStep("4", "TH sends TestEventTrigger command to General Diagnostics Cluster on Endpoint 0 with EnableKey field set to PIXIT.DEM.TEST_EVENT_TRIGGER_KEY and EventTrigger field set to PIXIT.DEM.TEST_EVENT_TRIGGER for Forecast Test Event Clear",
+                     "Verify DUT responds with status SUCCESS(0x00)"),
+        ]
+
+        return steps
+
+    @async_test_body
+    async def test_TC_DEM_2_9(self):
+        # pylint: disable=too-many-locals, too-many-statements
+        """Run the test steps."""
+        self.step("1")
+        # Commission DUT - already done
+
+        # Subscribe to Events and when they are sent push them to a queue for checking later
+        events_callback = EventChangeCallback(Clusters.DeviceEnergyManagement)
+        await events_callback.start(self.default_controller,
+                                    self.dut_node_id,
+                                    self.matter_test_config.endpoint)
+
+        self.step("2")
+        await self.check_test_event_triggers_enabled()
+
+        self.step("3")
+        await self.send_test_event_trigger_forecast()
+
+        self.step("3a")
+
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
+        asserts.assert_greater_equal(len(forecast.slots), 1)
+        asserts.assert_is_not_none(forecast.slots[0].manufacturerESAState)
+
+        self.step("3b")
+        forecast = await self.read_dem_attribute_expect_success(attribute="Forecast")
```

Comment:
There should be a PICS guard (only run if PICS_SF_PFR)

Reply:
Fixed

Reply:
I think this is ok now but this relies on the DUT telling us what it supports not what the PICS (from Test harness wants the DUT to support) - but I think there is a plan to move to this direction instead of relying on error prone PICS.